geode-core/src/main/java/org/apache/geode/cache/query/internal/index/CompactRangeIndex.java

```diff
@@ -803,8 +809,13 @@ private void addToResultsFromEntries(Object lowerBoundKey, Object upperBoundKey,
               if (runtimeItr != null) {
                 runtimeItr.setCurrent(value);
               }
-              if (ok && runtimeItr != null && iterOps != null) {
-                ok = QueryUtils.applyCondition(iterOps, context);
+              try {
+                if (ok && runtimeItr != null) {
+                  observer.beforeIterationEvaluation(iterOps, value);
+                  ok = QueryUtils.applyCondition(iterOps, context);
+                }
+              } finally {
+                observer.afterIterationEvaluation(ok);
```

Comment:
This class also has several try-finally blocks like that this may benefit by being extracted as a new private method both for readability and to hopefully reduce some of the repetition in these blocks.